### PR TITLE
`paginate_relations_with_filter` DRY & flip ordering #504

### PR DESCRIPTION
**Pull request by isosphere**
The two branches of the `match` pattern on `dir` used identical code. I removed the code duplication and reversed the order of the returned `chunk` - looks like we got it backwards the first time around.

Threaded messages load in the correct order in my Element Desktop instance with these changes.

### CODE REVIEW

**Comment by isosphere** [comment]
Looks like the "Documentation and GitHub Pages" action is broken, failed above. 

```
Run echo 'source $HOME/.nix-profile/share/nix-direnv/direnvrc' > "$HOME/.direnvrc"
  echo 'source $HOME/.nix-profile/share/nix-direnv/direnvrc' > "$HOME/.direnvrc"
  nix profile install --inputs-from . nixpkgs#direnv nixpkgs#nix-direnv
  direnv allow
  nix develop .#all-features --command true
  shell: /usr/bin/bash -e {0}
  env:
    SCCACHE_GHA_ENABLED: false
    RUSTC_WRAPPER: 
    SCCACHE_BUCKET: 
    SCCACHE_S3_USE_SSL: 
    SCCACHE_REGION: 
    SCCACHE_ENDPOINT: 
    SCCACHE_CACHE_MULTIARCH: 
    AWS_ACCESS_KEY_ID: 
    AWS_SECRET_ACCESS_KEY: 
    TERM: ansi
    ATTIC_TOKEN: 
    CACHIX_AUTH_TOKEN: 
    CARGO_INCREMENTAL: 0
    ATTIC_ENDPOINT: 
    ATTIC_PUBLIC_KEY: 
    NIX_CONFIG: show-trace = true
  extra-substituters = https://attic.kennel.juneis.dog/conduit https://attic.kennel.juneis.dog/conduwuit https://cache.lix.systems/ https://conduwuit.cachix.org/
  extra-trusted-public-keys = conduit:eEKoUwlQGDdYmAI/Q/0slVlegqh/QmAvQd7HBSm21Wk= conduwuit:BbycGUgTISsltcmH0qNjFR9dbrQNYgdIAcmViSGoVTE= cache.lix.systems:aBnZUw8zA7H35Cz2RyKFVs3H4PlGTLawyY5KRbvJR8o= conduwuit.cachix.org-1:MFRm6jcnfTf0jSAbmvLfhO3KBMt4px+1xaereWXp8Xg=
  
    CARGO_PROFILE: test
    NIX_PROFILES: /nix/var/nix/profiles/default /home/runner/.nix-profile
    NIX_USER_PROFILE_DIR: /nix/var/nix/profiles/per-user/runner
    NIX_SSL_CERT_FILE: /etc/ssl/certs/ca-certificates.crt
    NIX_USER_CONF_FILES: /tmp/tmp.qGrnx7dGVp/flake-nixConfig.conf:/home/runner/.config/nix/nix.conf:
error: path '/nix/store/d1xx41xdph2jhwj8wlwqdxzc8xzchf63-cachix-1.7.4-bin' is not valid
Error: Process completed with exit code 1.
```

**Comment by girlbossceo** [comment]
Seems like you need to run cargo fmt nightly.

```
       lints::cargo-fmt │O│ Diff in /home/runner/work/conduwuit/conduwuit/src/service/rooms/pdu_metadata/mod.rs:94:
      lints::cargo-fmt │O│  			1
      lints::cargo-fmt │O│  		};
      lints::cargo-fmt │O│  
      lints::cargo-fmt │O│ -				let relations_until = &self.relations_until(sender_user, room_id, target, from, depth)?;
      lints::cargo-fmt │O│ +		let relations_until = &self.relations_until(sender_user, room_id, target, from, depth)?;
      lints::cargo-fmt │O│  		let events: Vec<_> = relations_until // TODO: should be relations_after
      lints::cargo-fmt │O│                      .iter()
      lints::cargo-fmt │O│                      .filter(|(_, pdu)| {
      lints::cargo-fmt │O│ Diff in /home/runner/work/conduwuit/conduwuit/src/service/rooms/pdu_metadata/mod.rs:133:
      lints::cargo-fmt │O│  				.collect(),
      lints::cargo-fmt │O│  		};
      lints::cargo-fmt │O│  
      lints::cargo-fmt │O│ -				Ok(get_relating_events::v1::Response {
      lints::cargo-fmt │O│ +		Ok(get_relating_events::v1::Response {
      lints::cargo-fmt │O│  			chunk: events_chunk,
      lints::cargo-fmt │O│ -					next_batch: next_token.map(|t| t.stringify()),
      lints::cargo-fmt │O│ -					prev_batch: Some(from.stringify()),
      lints::cargo-fmt │O│ -					recursion_depth: if recurse {
      lints::cargo-fmt │O│ -						Some(depth.into())
      lints::cargo-fmt │O│ -					} else {
      lints::cargo-fmt │O│ -						None
      lints::cargo-fmt │O│ -					},
      lints::cargo-fmt │O│ -				})
      lints::cargo-fmt │O│ +			next_batch: next_token.map(|t| t.stringify()),
      lints::cargo-fmt │O│ +			prev_batch: Some(from.stringify()),
      lints::cargo-fmt │O│ +			recursion_depth: if recurse {
      lints::cargo-fmt │O│ +				Some(depth.into())
      lints::cargo-fmt │O│ +			} else {
      lints::cargo-fmt │O│ +				None
      lints::cargo-fmt │O│ +			},
      lints::cargo-fmt │O│ +		})
      lints::cargo-fmt │O│  	}
      lints::cargo-fmt │O│  
      lints::cargo-fmt │O│  	pub fn relations_until<'a>(
```

`cargo +nightly fmt`

**Comment by girlbossceo** [comment]
Also an unneeded late variable init:

```
  lints::clippy/default │E│     Checking conduit_service v0.4.6 (/home/runner/work/conduwuit/conduwuit/src/service)
 lints::clippy/default │E│ error: unneeded late initialization
 lints::clippy/default │E│    --> src/service/rooms/pdu_metadata/mod.rs:88:3
 lints::clippy/default │E│     |
 lints::clippy/default │E│ 88  |         let next_token;
 lints::clippy/default │E│     |         ^^^^^^^^^^^^^^^ created here
 lints::clippy/default │E│ ...
 lints::clippy/default │E│ 122 |         next_token = events.last().map(|(count, _)| count).copied();
 lints::clippy/default │E│     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ initialised here
 lints::clippy/default │E│     |
 lints::clippy/default │E│     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_late_init
 lints::clippy/default │E│     = note: `-D clippy::needless-late-init` implied by `-D warnings`
 lints::clippy/default │E│     = help: to override `-D warnings` add `#[allow(clippy::needless_late_init)]`
 lints::clippy/default │E│ help: move the declaration `next_token` here
 lints::clippy/default │E│     |
 lints::clippy/default │E│ 88  ~         
 lints::clippy/default │E│ 89  |
 lints::clippy/default │E│   ...
 lints::clippy/default │E│ 121 |
 lints::clippy/default │E│ 122 ~         let next_token = events.last().map(|(count, _)| count).copied();
 lints::clippy/default │E│     |
 lints::clippy/default │E│ 
 lints::clippy/default │E│ error: could not compile `conduit_service` (lib) due to 1 previous error
```

**Comment by isosphere** [comment]
> cargo +nightly fmt

This makes 214 changes to whitespace across the repo. I'll ignore all the changes it makes to other files. I can't tell the difference in my editor, looks like maybe a tab/spaces thing.

**Comment by girlbossceo** [comment]
Huh, that's strange. It shouldn't make that many. Try that yeah and we'll see what CI says, maybe nightly cargo-fmt made some changes.

**Comment by isosphere** [comment]
I will try to rebase to clean the commit log; currently stumbling with this step.

**Comment by girlbossceo** [comment]
Looks good, main CI job passed and is now just doing binary builds. Will merge when it passes.

File: src/service/rooms/pdu_metadata/mod.rs [hunk 1]
```diff
@@ -121,70 +117,30 @@ impl Service {
...
-										.map_or(true, |r| &content.relates_to.rel_type == r)
-								} else {
-									false
-								}
-						})
-					.take(limit)
-					.filter(|(_, pdu)| {
-						self.services
-							.state_accessor
-							.user_can_see_event(sender_user, room_id, &pdu.event_id)
```
Comment (girlbossceo) [inline src/service/rooms/pdu_metadata/mod.rs:163]: Does `user_can_see_event` get called anywhere else? This feels like a pretty important thing to have.

Comment (girlbossceo) [reply src/service/rooms/pdu_metadata/mod.rs:163]: Oops nevermind, just expanded the code above.

**Pull request merged**